### PR TITLE
enhance optimizer.py to better collect grads

### PR DESCRIPTION
**Pull request by ogvalt**
## Description

I've switch to check whereas state has attribute "model_grads" and eliminate transfering data from 
gpu to cpu which causes training slow down 

## Related Issue

#655 

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the [Code of Conduct](https://github.com/catalyst-team/catalyst/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I have read the [Contributing](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md) guide.
- [x] I have read I need to click 'Login as guest' to see Teamcity build logs
- [x] I have checked the code-style using `make check-codestyle`.
- [x] I have written the docstring in Google format for all the methods and classes that I used.
- [ ] I have checked the docs using `make check-docs`.

### CODE REVIEW

File: catalyst/core/callbacks/optimizer.py [hunk 1]
```diff
@@ -173,10 +168,19 @@ def on_batch_end(self, state: _State):
                 grad_clip_fn=self.grad_clip_fn
             )
 
-            if self.save_model_grads:
-                for tag, value in model.named_parameters():
-                    tag = tag.replace(".", "/")
-                    state.model_grads[tag] = value.grad.cpu().numpy()
+            if hasattr(state, "model_grads"):
+                memo = set()
```
Comment (Scitator) [inline catalyst/core/callbacks/optimizer.py:172]: I dont like the idea of using `hasattr ` logic with State in Callbacks.
It looks like undocumented trick. Too many magic.

**Review by Scitator — CHANGES_REQUESTED** [review]

File: catalyst/core/callbacks/optimizer.py (hunk 1 above)
Comment (ogvalt) [reply catalyst/core/callbacks/optimizer.py:172]: Okay. Previously I've found that when I suggested adding model_grads to RunnerState ( https://github.com/catalyst-team/catalyst/pull/581/commits/5b99adeca617d8f0b74703649d6c6b377f06e08b ) 
the changes wasn't merged. So right know it's still remains undocumented feature. I could add them again in this PR and makes everything more convenient

**Comment by Scitator** [comment]
could you please merge with master? we have tests update

**Comment by Scitator** [comment]
closed due to inactivity :)

**Pull request closed without merging**